File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/com/ririnto/sinon/harness/rules/text/DocHeadingsRule.java
package com.ririnto.sinon.harness.rules.text;

import com.ririnto.sinon.harness.rules.HarnessCheckHelper;
import com.ririnto.sinon.harness.rules.HarnessCheckRule;
import com.ririnto.sinon.harness.core.RuleContext;
import com.ririnto.sinon.harness.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;

/**
 * Rule that requires specific headings in documentation files.
 */
public enum DocHeadingsRule implements HarnessCheckRule {
    INSTANCE;

    private static final String CATEGORY = "docHeadings";

    @Override
    public String category() {
        return CATEGORY;
    }

    @Override
    public boolean applies(RuleContext ctx) {
        return ctx.manifest().isEnabled(CATEGORY);
    }

    @Override
    public Collection<Finding> validate(RuleContext ctx) throws MojoExecutionException {
        final Path root = ctx.root();
        final JsonNode manifest = ctx.manifest().raw();
        final JsonNode catNode = manifest.get(CATEGORY);
        final JsonNode parameters = catNode.get("parameters");
        final JsonNode sourceFilter = parameters.get("sourceFilter");
        final String prefix = sourceFilter.get("prefix").asText();
        final String suffix = sourceFilter.get("suffix").asText();
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        final List<String> headings = HarnessCheckHelper.extractPaths(parameters.get("headings"));
        return HarnessCheckHelper.extractPaths(manifest.get("filePresence").get("parameters").get("paths")).stream()
                .filter(f -> f.startsWith(prefix) && f.endsWith(suffix))
                .flatMap(f -> validateHeadings(root, f, headings, severity).stream())
                .toList();
    }

    private List<Finding> validateHeadings(Path root, String file, List<String> headings, String severity) {
        final Path filePath = root.resolve(file);
        if (!HarnessCheckHelper.isSafeRegularFile(root, filePath)) {
            return List.of();
        }
        final String text;
        try {
            text = HarnessCheckHelper.readFile(root, filePath);
        } catch (MojoExecutionException e) {
            return List.of();
        }
        return headings.stream()
                .filter(h -> !text.contains(h))
                .map(h -> Finding.of(severity, CATEGORY, "doc missing " + h + ": " + file))
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/com/ririnto/sinon/harness/rules/text/DocHeadingsRule.kt
package com.ririnto.sinon.harness.rules.text

import com.ririnto.sinon.harness.ast.HarnessAstResults.Finding
import com.ririnto.sinon.harness.core.JsonAccess
import com.ririnto.sinon.harness.core.RuleContext
import com.ririnto.sinon.harness.rules.HarnessCheckRule
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.commonmark.node.AbstractVisitor
import org.commonmark.node.HardLineBreak
import org.commonmark.node.Heading
import org.commonmark.node.Node
import org.commonmark.node.SoftLineBreak
import org.commonmark.node.Text
import org.commonmark.parser.Parser
import java.nio.file.Path

/**
 * Rule that requires documentation files to contain specified headings.
 *
 * @category docHeadings
 */
object DocHeadingsRule : HarnessCheckRule() {
    /**
     * Category identifier for this rule.
     */
    override val category: String = "docHeadings"
    private val markdownParser: Parser = Parser.builder().build()

    override fun validate(ctx: RuleContext): Collection<Finding> {
        val parametersObj =
            ctx.manifest
                .categoryObject(category)
                ?.get("parameters")
                ?.jsonObject ?: return emptyList()
        val sourceFilterObj = parametersObj["sourceFilter"]?.jsonObject
        return JsonAccess
            .stringArrayFromObject(
                referencedCategoryParameters(ctx, parametersObj),
                "paths",
            ).filter { sourceFile -> sourceFile.startsWith(JsonAccess.stringFromObject(sourceFilterObj, "prefix")) }
            .filter { sourceFile -> sourceFile.endsWith(JsonAccess.stringFromObject(sourceFilterObj, "suffix")) }
            .flatMap { docPath ->
                ctx.manifest
                    .stringArray(category, "headings")
                    .filter { heading -> !extractHeadings(markdownParser.parse(ctx.readSafe(docPath))).contains(heading) }
                    .map { heading ->
                        Finding(
                            ctx.manifest.severityOf(category),
                            category,
                            "doc missing $heading: $docPath",
                        )
                    }
            }
    }

    /**
     * Retrieves the parameters object referenced by the sourceFilesFromCategory field.
     *
     * @param ctx Rule context containing manifest data.
     * @param parametersObj Parameters object with sourceFilesFromCategory field.
     * @return Referenced parameters object or null if not found.
     */
    private fun referencedCategoryParameters(
        ctx: RuleContext,
        parametersObj: JsonObject,
    ): JsonObject? {
        return ctx.manifest.raw[JsonAccess.stringFromObject(parametersObj, "sourceFilesFromCategory")]
            ?.jsonObject
            ?.get("parameters")
            ?.jsonObject
    }

    /**
     * Extracts all heading text content from a document node tree.
     *
     * @param node Root node to traverse.
     * @return Set of heading text strings found in the node.
     */
    private fun extractHeadings(node: Node): Set<String> =
        buildSet {
            node.accept(HeadingsCollector(this::add))
        }

    /**
     * Converts all children of this node to a string, joining text and line breaks.
     *
     * @return String representation of node children.
     */
    private fun Node.childrenToString(): String =
        generateSequence(firstChild) { current -> current.next }
            .map { child ->
                when (child) {
                    is Text -> child.literal
                    is SoftLineBreak -> " "
                    is HardLineBreak -> " "
                    else -> ""
                }
            }.joinToString("")

    /**
     * Visitor for extracting heading text from Markdown nodes.
     */
    private class HeadingsCollector(
        private val record: (String) -> Unit,
    ) : AbstractVisitor() {
        override fun visit(heading: Heading) {
            val text = heading.childrenToString()
            if (text.isNotEmpty()) {
                record(text)
            }
        }

        /**
         * Converts all children of this node to a string, joining text and line breaks.
         */
        private fun Node.childrenToString(): String =
            generateSequence(firstChild) { current -> current.next }
                .map { child ->
                    when (child) {
                        is Text -> child.literal
                        is SoftLineBreak -> " "
                        is HardLineBreak -> " "
                        else -> ""
                    }
                }.joinToString("")
    }
}
